src/m3/constraints.ts

```diff
@@ -12,19 +19,37 @@ export const issuesMetamodel = (metamodel: Metamodel): Issue[] =>
     flatMap(
         metamodel,
         (t) => {
+
+            const issues: Issue[] = []
+            const issue = (message: string): void => {
+                issues.push({
+                    location: t,
+                    message
+                })
+            }
+
+            if (t instanceof Concept || t instanceof ConceptInterface) {
+                const cycle = inheritedCycleWith(t)
+                if (cycle.length > 0) {
+                    issue(`A ${t.constructor.name} can't inherit (directly or indirectly) from itself, but ${t.qualifiedName()} does so through the following cycle: ${cycle.map((t) => t.qualifiedName()).join(" -> ")}`)
+                }
+            }
+
             if (t instanceof ConceptInterface) {
                 const nonDerivedFeatures = t.allFeatures().filter(({derived}) => !derived)
                 if (nonDerivedFeatures.length > 0) {
                     const isPlural = nonDerivedFeatures.length > 1
-                    return [
-                        {
-                            location: t,
-                            message: `The features of a ConceptInterface must all be derived, but the following feature${isPlural ? `s` : ``} of ${t.simpleName} ${isPlural ? `are` : `is`} not: ${nonDerivedFeatures.map(({simpleName}) => simpleName).join(", ")}.`
-                        }
-                    ]
+                    issue(`The features of a ConceptInterface must all be derived, but the following feature${isPlural ? `s` : ``} of ${t.qualifiedName()} ${isPlural ? `are` : `is`} not: ${nonDerivedFeatures.map(({simpleName}) => simpleName).join(", ")}.`)
                 }
             }
-            return []
+
+            if (t instanceof Link) {
+                if (t.type instanceof Annotation) {
```

Comment:
This makes me wonder if it should be renamed in t.type, as these two should probably have the same name:

```
Annotation "*" -- "1" AbstractConcept: target
Link "*" -- "1" AbstractConcept: type
```

Reply:
Given that we're going to remove annotations anyway (for now), I guess this problem disappears into thin air :)